### DISCUSSION TITLE
Which store properties to return

### DISCUSSION THREAD

**Discussion by jdnz** [vuejs/pinia#2402] [Help and Questions] [answered]
I am transitioning to using setup stores in my project and I am confused about what the following statement in the documentation means:

> Note you must return all state properties in setup stores for pinia to pick them up as state. In other words, you cannot have private state properties in stores.

What does it mean for pinia to  'pick them up as state'? More specifically I'm wondering if for some reason I have to return everything from the store, i.e. all the refs, computeds and functions for them to work correctly (trigger reactivity etc), even if they are only used internally in the store. Or do I only have to return things that I actually use outside the store?

**Comment by gegoune** [comment]
The latter. You don't need to return anything that you don't want to directly access outside your store.

**Comment by posva** [comment, marked as answer]
You need to return all `ref()`, `reactive()`, `shallowReactive()`. Anything that is state. You only need to return functions and computed that are used outside of the store.

**Reply by gunhaxxor** [reply]
This is still not 100% clear to me. Just to make sure, if I have a ref inside a setup store and I have a computed that is derived from that ref, I still need to return the ref, even if I will _never_ use it outside the store?
If this is the case, why? What will break and how will it break?

As of now I believe I have the given scenario in some of my stores and I'm worried it will lead to bugs I'm unaware of.

**Reply by posva** [reply]
Yes, you have to return it but you can actually hide it [like this](https://masteringpinia.com/blog/my-top-5-tips-for-using-pinia#private-state-with-nested-stores). I do talk more about the problem within the course but the short version is: breaks SSR, devtools, and some plugins

**Reply by gunhaxxor** [reply]
Thanks for the quick reply. Good to know!

**Reply by paro-paro** [reply]
Hi @posva 

The [dealing with composables (setup stores)](https://pinia.vuejs.org/cookbook/composables.html#Setup-Stores) section from the docs shows a `videoPlayer` ref that is not returned from the store with a explicit `we won't expose this element directly` comment on top.

I am still wrapping my head around composition/composables, stores, etc, in vue 3 so any help to clarify this use case will  really help. 

Thanks!

**Reply by posva** [reply]
The ref of an HTMLElement is quite the edge case, so it's normal this can be confusing. I added a note to that page of the docs. I would recommend you not to focus too much on that specific example.

**Reply by paro-paro** [reply]
Thanks for the clarification! Must appreciated

**Reply by zumm** [reply]
@posva Hi. Im confused with this concept too. Could you please elaborate how does it work with composables? Pinia documantation claims we can use almost any composable with setup store function. But composables may have internal state. Why is it ok in this case?
